hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyRackFaultTolerantWithExcludedScope.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.blockmanagement;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hdfs.protocol.DatanodeInfo;
+import org.apache.hadoop.net.NetworkTopology;
+import org.apache.hadoop.net.NetworkTopologyWithExcludedScope;
+import org.apache.hadoop.service.ServiceStateException;
+
+/**
+ * The class is responsible for choosing the desired number of targets
+ * for placing block replicas.
+ * The strategy is that it tries its best to place the replicas to most racks.
+ */
+@InterfaceAudience.Private
+public class BlockPlacementPolicyRackFaultTolerantWithExcludedScope extends BlockPlacementPolicyRackFaultTolerant {
+  
+  @Override
+  public void initialize(Configuration conf, FSClusterStats stats, NetworkTopology clusterMap,
+                         Host2NodesMap host2datanodeMap) {
+    if (!(clusterMap instanceof NetworkTopologyWithExcludedScope)) {
+      throw new ServiceStateException(
+          "BlockPlacementPolicyRackFaultTolerantWithExcludedRacks requires the use of NetworkTopologyWithExcludedScope"
+      );
+    }
+    super.initialize(conf, stats, clusterMap, host2datanodeMap);
+  }
+  
+  // Same as in BlockPlacementPolicyRackFaultTolerant but checks if the node is from the excluded scope
+  // It was not possible to factorise with BlockPlacementPolicyRackFaultTolerant
+  @Override
+  public BlockPlacementStatus verifyBlockPlacement(DatanodeInfo[] locs,
+      int numberOfReplicas) {
+    if (locs == null)
+      locs = DatanodeDescriptor.EMPTY_ARRAY;
+    if (!clusterMap.hasClusterEverBeenMultiRack()) {
+      // only one rack
+      return new BlockPlacementStatusDefault(1, 1, 1);
+    }
+    // 1. Check that all locations are different.
+    // 2. Count locations on different racks.
+    // Make sure excluded racks are not accounted
+    Set<String> racks = new TreeSet<>();
```

Comment:
Any reason to use `TreeSet` instead of `HashSet` as in `BlockPlacementPolicyRackFaultTolerant.verifyBlockPlacement`?

Reply:
To be honest, I used the exact same as in the parent class, with just a little modification. I think for very small Sets, TreeSet might be more efficient since no hash is computed.